runtime/io.c

```diff
@@ -446,7 +499,12 @@ void caml_finalize_channel(value vchan)
 {
   struct channel * chan = Channel(vchan);
   if ((chan->flags & CHANNEL_FLAG_MANAGED_BY_GC) == 0) return;
-  if (--chan->refcount > 0) return;
+  unlink_channel(chan);
+  if (atomic_fetch_add (&chan->refcount, -1) > 1) {
+    /* [caml_ml_out_channels_list] may have a reference to this channel. */
+    link_channel (chan);
+    return;
+  }
```

Comment:
Same comment as above.  Rather than "let's unlink and ooops let's link again", could we just unlink only when it's safe?